src/engine/renderer/tr_shader.cpp

```diff
@@ -4652,6 +4792,37 @@ static void CollapseStages()
 		{
 			stage->bundle[ TB_GLOWMAP ].image[ 0 ] = tr.blackImage;
 		}
+
+		// Compute normal scale.
+		for ( int i = 0; i < 3; i++ )
+		{
+			if ( stage->hasNormalMap )
+			{
+				if ( !stage->hasNormalFormat )
+				{
+					// Please make sure the parser sets a normal format
+					// when adding a new normal map syntax.
+					ASSERT_UNREACHABLE();
+				}
+
+				if ( !stage->hasNormalScale )
+				{
+					stage->normalScale[ i ] = 1.0;
+				}
+
+				// If height map is not stored in normal map alpha channel,
+				// the Z component will be reconstructed and will be fine
+				// from the start, so we must not translate the Z format.
+				if ( i < 2 || stage->isHeightMapInNormalMap )
```

Comment:
What is the idea of checking isHeightMapInNormalmap?

Reply:
That's related to the fact Z reconstruction is never done when normal map has heightmap in normal map (as reconstruction is destructive on alpha channel). This code applies format translation for X and Y in all case, and Z only when it's not reconstructed. I will make the comment more explicit.